python_uds_mqttt/mqtt_udsource.py

```diff
@@ -0,0 +1,159 @@
+    """
+    User-defined source for MQTT messages.
+    """
+
+    def __init__(self, broker, port, topic):
+        # The offset idx till where the messages have been read
+        self.read_idx: int = 0
+        # Set to maintain a track of the offsets yet to be acknowledged
+        self.to_ack_set: set[int] = set()
+        # Set to maintain a track of the offsets that have been negatively acknowledged
+        self.nacked: set[int] = set()
+        # MQTT broker address to connect to.
+        self.broker = broker
+        # Port number of the MQTT broker.
+        self.port = port
+        # MQTT topic to subscribe to for receiving messages.
+        self.topic = topic
+        # Async queue to store incoming messages
+        self.messages = asyncio.Queue()
+        # Asyncio task for MQTT client loop
+        self._mqtt_task = None
+        # Flag indicating if source has started
+        self._started = False
+
+    async def start_mqtt_consumer(self):
+        """Start the MQTT consumer"""
+        if self._started:
+            return
+        self._started = True
+        
+        logger.info(f"Starting MQTT consumer for broker={self.broker}, port={self.port}, topic={self.topic}")
+        
+        async def mqtt_loop():
+            while True:
+                try:
+                    async with Client(self.broker, self.port) as client:
+                        await client.subscribe(self.topic)
+                        logger.info(f"Successfully subscribed to MQTT topic: {self.topic}")
+                        async for msg in client.messages:
+                            payload = msg.payload.decode()
+                            logger.info(f"Received MQTT message: {payload}")
+                            await self.messages.put(payload)
+                except Exception as e:
+                    logger.error(f"MQTT consumer error: {e}. Retrying in 5 seconds...")
+                    await asyncio.sleep(5)
+        
+        self._mqtt_task = asyncio.create_task(mqtt_loop())
+
+    async def read_handler(self, datum: ReadRequest, output: NonBlockingIterator):
+        """
+        read_handler is used to read the data from the source and send the data forward
+        for each read request we process num_records and increment the read_idx to indicate that
+        the message has been read and the same is added to the ack set
+        """
+
+        if not self._started:
+            await self.start_mqtt_consumer()
+        
+        if len(self.to_ack_set) >= 500:
+            return
+
+        for _ in range(datum.num_records):
+            if self.nacked:
+                idx = self.nacked.pop()
+            else:
+                idx = self.read_idx
+                self.read_idx += 1
+
+            try:
+                payload = self.messages.get_nowait()
+                logger.info(f"Sending MQTT message: {payload}")
+            except asyncio.QueueEmpty:
+                payload = f"dummy-{idx}"
```

Comment:
This will create dummy payloads which is not right. You do not have to return anything if there is no data. In the README, just add how you can send some data to MQTT broker via command-line.


```
{"timestamp":"2025-11-17T00:23:16.613066Z","level":"INFO","message":"Payload - dummy-669498 Keys -  EventTime - 1763338996579 Headers - x-txn-id: 9cac40fb-4c59-4df1-ac23-d77128c2105d,  ID - log-sink-669499-0-0","target":"numaflow_core::sinker::sink::log"}
{"timestamp":"2025-11-17T00:23:16.613068Z","level":"INFO","message":"Payload - dummy-669499 Keys -  EventTime - 1763338996579 Headers - x-txn-id: 6a429484-9fb6-4342-ad72-d598499f413f,  ID - log-sink-669500-0-0","target":"numaflow_core::sinker::sink::log"}
```